mysite/lti/models.py

```diff
@@ -0,0 +1,94 @@
+import json
+from django.db import models
+from django.contrib.auth import login
+from django.contrib.auth.models import User
+
+from social.apps.django_app.default.models import UserSocialAuth
+
+from ct.models import Role, Course
+
+
+class LTIUser(models.Model):
+    """Model for LTI user
+
+    Created for Moodle LMS.
+    Intended to link to Django user in socraticqs2.
+    """
+    user_id = models.CharField(max_length=255, blank=False)
+    consumer = models.CharField(max_length=64, blank=True)
+    extra_data = models.TextField(max_length=1024, blank=False)
+    django_user = models.ForeignKey(User, null=True, related_name='lti_auth')
+    course_id = models.IntegerField()
+
+    class Meta:
+        unique_together = ('user_id', 'consumer', 'course_id')
+
+    def create_links(self):
+        """Create all needed links to Django and/or UserSocialAuth"""
+        extra_data = json.loads(self.extra_data)
+        username = extra_data.get('lis_person_name_full', self.user_id)
+        first_name = extra_data.get('lis_person_name_given', '')
+        last_name = extra_data.get('lis_person_name_family', '')
+        email = extra_data.get('lis_person_contact_email_primary', '').lower()
+
+        if email:
+            social = UserSocialAuth.objects.filter(provider='email',
+                                                   uid=email)
+            if social:
+                django_user = social[0].user
+            else:
+                django_user = User.objects.filter(email=email)
```

Comment:
Django 1.6 (released Nov 2013) introduced the convenience methods first() and last() which swallow the resulting exception and return None if the queryset returns no objects. (http://stackoverflow.com/questions/5123839/fastest-way-to-get-the-first-object-from-a-queryset-in-django). See https://docs.djangoproject.com/en/dev/ref/models/querysets/#first


So the whole code can look like
```
defaults = {
     'first_name': first_name,
     'last_name': last_name,
}

if email:
    defaults['email'] = email
    social = UserSocialAuth.objects.filter(provider='email', uid=email).first()

    if social:
        django_user = social.user
    else:
        django_user = User.objects.filter(email=email).first()

        if not django_user:
            django_user, created = User.objects.get_or_create(
                username=username, defaults=defaults
            )

        social = UserSocialAuth(user=django_user,
                                provider='email',
                                uid=email,
                                extra_data=extra_data)
        social.save()
else:
    django_user, created = User.objects.get_or_create(username=username, defaults=defaults
    )

self.django_user = django_user
self.save()


```